OBSERVE-2845 Implement various fixes from live stage testing

arjunnayak opened the pull request:
- Add health check extension to custom base otel collector build (it's used in our k8s deployment)
- Build amd64 binary for running on k8s nodes (the vast majority of our apps/infra run on linux amd64)
- Fixes processor code to handle live data in stage: response_code instead of status_code, endpoint type handling
- Added debug logs for easier debugging of live incoming log records

arjunnayak committed:
Add health check extension to otel collector–used in k8s

arjunnayak committed:
Build amd64 binary for running in k8s

arjunnayak committed:
Update dockerfile–add test slo.yaml to container, document expose port for metrics

arjunnayak committed:
Fix processor to successfully handle live data in stage

- correct status code field to response_code
- fixes endpoint type handling, it should default to http
- debug log records

arjunnayak committed:
Bump version to 0.0.13

arjunnayak committed:
clean up slo.yaml

benjdauno commented on Dockerfile:
```diff
@@ -9,6 +9,7 @@ USER ${USER_UID}:${USER_GID}
 COPY --from=prep /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
 COPY bin/affirm-otelcol /
-EXPOSE 4317 55680 55679
+COPY dev_tooling/slo.yaml /
```
Can you volume mount this in the docker-compose file and k8s manifests instead?

benjdauno replied:
Unless we're supplying a sane default.

arjunnayak replied:
should have clarified with comments–this is a temporary solution until we have the slo.yaml file in an s3 bucket that we pull initially and poll periodically for updates. Rohit is working on that in the next 2 weeks, and I will direct him to remove this code

Aloshi commented on slometricsemitter/log_processor.go:
```diff
@@ -198,20 +201,24 @@ func (sloMetricsProc *sloMetricsProcessor) getAttributeFromLogRecord(logRecord p
 }
 
-func (sloMetricsProc *sloMetricsProcessor) determineEndpointType(logRecord plog.LogRecord) (string, error) {
+// Defaults to http if content_type attribute is missing or unknown.
```
Is there a reason you chose to default to `http` here instead of some `unknown` value? is this a common situation?

arjunnayak replied:
i actually just copied what we do in [latency reporter](https://github.com/Affirm/all-the-things/blob/master/jvm-o11y-kt/latency-reporting-flink/src/main/kotlin/aff/o11yKt/latencyReporting/common/EnvoyAccessLogDeserializer.kt#L90-L94)

arjunnayak replied:
i'll look into this and when it actually happens

arjunnayak replied:
based on testing in stage, content-type just isn't strongly set in http responses so we need to leave this behavior in to capture as comprehensive of a list of endpoints as possible

Aloshi reviewed (approved):
seems reasonable to me

arjunnayak committed:
add note to dockerfile

arjunnayak committed:
Move istio log extraction to a separate method with a new struct. Separate telemetry setup. Move some method declarations around to follow a vertical hierarchy

arjunnayak committed:
Update endpoint type detector to return error if no content type is present instead of defaulting to http

arjunnayak committed:
Update endpoint detection when endpoint type is unknown

- the processor will try to get the endpoint name even with an unknown endpoint type
- this way it can still observe the overall request metric, but the latency breach metric will be skipped

arjunnayak committed:
Update endpoint type handling to default to http.

- After testing, it appears content-type isn't strongly set in our http responses.
- It's better for us to try to capture metrics for those endpoints than missing them

arjunnayak committed:
bump to 0.0.14

arjunnayak committed:
rm noisy log, bump to v0.0.15
